### PR TITLE
More robust JSON parsing

### PR DESCRIPTION
**Pull request by whitead**
* Added more json variants observed from Claude
* Added unit test with pytest features
* Removed older asyncio unit testing syntax
* Incremented version

### CODE REVIEW

File: paperqa/utils.py [hunk 1]
```diff
@@ -179,3 +180,16 @@ def is_coroutine_callable(obj):
...
+def llm_read_json(text: str) -> dict:
+    """Read LLM output and extract JSON data from it."""
+    # fetch from markdown ```json if present
+    text = text.strip().split("```json")[-1].split("```")[0]
+    print(text)
+    # split anything before the first {
+    text = "{" + text.split("{", 1)[-1]
+    # split anything after the last }
+    text = text.rsplit("}", 1)[0] + "}"
+    print(text)
```
Comment (jamesbraza) [inline paperqa/utils.py:194]: Extra `print`s here?

File: tests/test_paperqa.py [hunk 2]
```diff
@@ -409,101 +409,137 @@ def test_extract_score():
...
+    def accum(x):
+        outputs.append(x)
+
+    outputs: list[str] = []
+    completion = await call({"animal": "duck"}, callbacks=[accum])  # type: ignore[call-arg]
+    assert completion.seconds_to_first_token > 0
+    assert completion.prompt_count > 0
+    assert completion.completion_count > 0
+    assert str(completion) == "".join(outputs)
+    assert type(completion.text) is str  # noqa: E721
```
Comment (jamesbraza) [inline tests/test_paperqa.py:526]:
```suggestion
    assert isinstance(completion.text, str)
```

And a few lines below

File: tests/test_paperqa.py [hunk 3]
```diff
@@ -948,126 +984,124 @@ def test_langchain_embeddings():
...
-        index = LangchainVectorStore(cls=FAISS, embedding_model=OpenAIEmbeddings())
-        assert index._store is None
+    ]
+
+    # checks on builder
+    try:
+        index = LangchainVectorStore()
         index.add_texts_and_embeddings(some_texts)
-        assert index._store is not None
+        raise "Failed to check for builder"  # type: ignore[misc]  # noqa: B016
```
Comment (jamesbraza) [inline tests/test_paperqa.py:1005]: `raise "Failed to check for builder"`

Did you forget an `Exception` type here? I don't think this is really valid:

```python
>>> raise "hi"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: exceptions must derive from BaseException
```

File: tests/test_paperqa.py [hunk 4]
```diff
@@ -948,126 +984,124 @@ def test_langchain_embeddings():
...
+    except ValueError:
+        pass
```
Comment (jamesbraza) [inline tests/test_paperqa.py:1007]: For patterns like this, I like `with pytest.raises(ValueError)`

Comment (whitead) [reply tests/test_paperqa.py]: Yeah, this PR switches to pytest. Thanks for catching - will switch to pytest.

File: tests/test_paperqa.py (hunk 3 above)
Comment (whitead) [reply tests/test_paperqa.py]: Fixed!

File: tests/test_paperqa.py [hunk 5]
```diff
@@ -409,101 +409,137 @@ def test_extract_score():
...
+        client,
+        "The {animal} says",
+        skip_system=True,
+    )
+    outputs = []
+
+    def accum(x):
+        outputs.append(x)
+
+    completion = await call({"animal": "duck"}, callbacks=[accum])  # type: ignore[call-arg]
```
Comment (jamesbraza) [inline tests/test_paperqa.py:485]:
```suggestion
    outputs = []
    completion = await call({"animal": "duck"}, callbacks=accum.append)  # type: ignore[call-arg]
```

File: tests/test_paperqa.py [hunk 6]
```diff
@@ -409,101 +409,137 @@ def test_extract_score():
...
+    assert completion.seconds_to_first_token > 0
+    assert completion.prompt_count > 0
+    assert completion.completion_count > 0
+    assert str(completion) == "".join(outputs)
+    assert isinstance(completion.text, str)
+
+    completion = await call({"animal": "duck"})  # type: ignore[call-arg]
+    assert completion.seconds_to_first_token == 0
+    assert completion.seconds_to_last_token > 0
+    assert type(completion.text) is str  # noqa: E721
```
Comment (jamesbraza) [inline tests/test_paperqa.py:531]:
```suggestion
    assert isinstance(completion.text, str)
```

File: tests/test_paperqa.py [hunk 7]
```diff
@@ -948,126 +984,114 @@ def test_langchain_embeddings():
...
         index.add_texts_and_embeddings(some_texts)
-        assert index._store is not None
 
-        docs = Docs(
-            texts_index=LangchainVectorStore(
-                cls=FAISS, embedding_model=OpenAIEmbeddings()
-            )
-        )
-        assert docs._embedding_client is not None  # from docs_index default
+    with pytest.raises(ValueError):  # noqa: PT011
```
Comment (jamesbraza) [inline tests/test_paperqa.py:1005]: FYI you can do `pytest.raises(ValueError, match="some aspect of message")` for even more precise assertion

File: tests/test_paperqa.py [hunk 8]
```diff
@@ -948,126 +984,114 @@ def test_langchain_embeddings():
...
+    index = LangchainVectorStore(
+        store_builder=lambda x, y: FAISS.from_embeddings(x, OpenAIEmbeddings(), y)
+    )
+    assert index._store is None
+    index.add_texts_and_embeddings(some_texts)
+    assert index._store is not None
+    # check search returns Text obj
+    data, score = await index.similarity_search(None, "test", k=1)  # type: ignore[unreachable]
+    print(data)
+    assert type(data[0]) == Text
```
Comment (jamesbraza) [inline tests/test_paperqa.py:1021]:
```suggestion
    assert isinstance(data[0], Text)
```

**Review by jamesbraza — APPROVED** [review]
Added some more comments

**Pull request merged**